tests/test_request_fingerprinter.py

```diff
@@ -0,0 +1,468 @@
+# available, that could technically not be the case given a custom user
+# provider.
+FINGERPRINTING_INPUTS = (
+    BrowserHtml,
+    BrowserResponse,
+    HttpResponse,
+    HttpResponseBody,
+    HttpResponseHeaders,
+    ResponseUrl,
+)
+
+
+@pytest.mark.parametrize("input_cls", FINGERPRINTING_INPUTS)
+def test_fingerprinting_unannotated_page_inputs(input_cls):
+    """Inputs that may have an impact on the actual request sent even without
+    annotations."""
+
+    class TestSpider(Spider):
+        name = "test_spider"
+
+        async def parse_input(self, response, some_input: input_cls):
+            pass
+
+    crawler = get_crawler(spider_cls=TestSpider, ensure_providers_for=[input_cls])
+    fingerprinter = crawler.request_fingerprinter
+    request1 = Request("https://toscrape.com")
+    fingerprint1 = fingerprinter.fingerprint(request1)
+    request2 = Request("https://toscrape.com", callback=crawler.spider.parse_input)
+    fingerprint2 = fingerprinter.fingerprint(request2)
+    assert fingerprint1 != fingerprint2
+
+
+@pytest.mark.parametrize(
+    "input_cls_a, input_cls_b",
+    (tuple(combination) for combination in combinations(FINGERPRINTING_INPUTS, 2)),
+)
+def test_fingerprinting_unannotated_page_input_combinations(input_cls_a, input_cls_b):
+    """Make sure that a combination of known inputs that alter the request
+    fingerprint does not result in the same fingerprint."""
+
+    class TestSpider(Spider):
+        name = "test_spider"
+
+        async def parse_a(self, response, input_a: input_cls_a):
+            pass
+
+        async def parse_b(self, response, input_b: input_cls_b):
+            pass
+
+        async def parse_all(self, response, input_a: input_cls_a, input_b: input_cls_b):
+            pass
+
+    crawler = get_crawler(
+        spider_cls=TestSpider, ensure_providers_for=[input_cls_a, input_cls_b]
+    )
+    fingerprinter = crawler.request_fingerprinter
+    request1 = Request("https://toscrape.com", callback=crawler.spider.parse_a)
+    fingerprint1 = fingerprinter.fingerprint(request1)
+    request2 = Request("https://toscrape.com", callback=crawler.spider.parse_b)
+    fingerprint2 = fingerprinter.fingerprint(request2)
+    request3 = Request("https://toscrape.com", callback=crawler.spider.parse_all)
+    fingerprint3 = fingerprinter.fingerprint(request3)
+    assert fingerprint1 != fingerprint2
+    assert fingerprint1 != fingerprint3
+    assert fingerprint2 != fingerprint3
+
+
+def test_dep_resolution():
+    """Currently we do not resolve dependencies, so it is possible to get a
+    different fingerprint for callbacks that resolve to identical
+    dependencies.
+
+    The reason for not resolving dependencies is that it could be hard where
+    items are involved. This might be addressed in the future, in which case
+    we should consider to fingerprint based on leaf dependencies (those with a
+    provider) and not on the root dependencies present in the callback.
```

Comment:
I think we can agree that resolving dependencies now would be too hard given ItemProvider. But I think the question then becomes: do we keep this PR on hold until we can easily resolve dependencies, or do we merge without dependency resolving and then later on change all fingerprints (invalidating caches)?